### PR TITLE
Feature/clusters benchmarking

### PR DESCRIPTION
**Pull request by Raid5594**
- benchmarks for clusters
- weights for clusters
- extended node_provider_auth
- create staker

### CODE REVIEW

**Review by aie0 — APPROVED** [review]

File: pallets/ddc-clusters/src/node_provider_auth.rs [hunk 1]
```diff
@@ -52,19 +54,87 @@ where
...
+	}
+
 	pub fn new(contract_id: T::AccountId, caller_id: T::AccountId) -> Self {
 		Self { contract_id, caller_id }
 	}
 }
 
 pub enum NodeProviderAuthContractError {
-	ContractCallFailed,
+	ContractCall,
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/node_provider_auth.rs:137]: I would leave it as `ContractCallFailed`, and add `ContractDeployFailed`, `NodeAuthorizationUnsucceeded` to make error type more descriptive

**Review by yahortsaryk — CHANGES_REQUESTED** [review]

File: pallets/ddc-staking/src/lib.rs [hunk 2]
```diff
@@ -965,6 +965,34 @@ pub mod pallet {
...
+			Self::deposit_event(Event::<T>::Bonded(stash.clone(), value));
+			let item = StakingLedger {
+				stash: stash.clone(),
+				total: value,
+				active: value,
+				chilling: Default::default(),
+				unlocking: Default::default(),
+			};
+			Self::update_ledger(&controller, &item);
+			Self::do_add_cdn(&stash, cluster_id);
```
Comment (yahortsaryk) [inline pallets/ddc-staking/src/lib.rs:990]: As `bond_stake_and_serve` function accepts `NodePubKey` of any type (both Storage and CDN), it should be able to process a node of any type. Please add code that handles staking for the Storage node as well

**Review by yahortsaryk — CHANGES_REQUESTED** [review]

File: pallets/ddc-staking/src/lib.rs (hunk 2 above)
Comment (aie0) [reply pallets/ddc-staking/src/lib.rs:990]: it seems like we're constantly having these if/else for Storage/CDN. for now, let's fix it with another if/else, but once we have another node type, let's create a trait and factory returning different implementations for each node type

File: pallets/ddc-clusters/src/node_provider_auth.rs (hunk 1 above)
Comment (Raid5594) [reply pallets/ddc-clusters/src/node_provider_auth.rs:137]: I got an error from linter, as suffix was the same, but will find a way around

Comment (Raid5594) [reply pallets/ddc-clusters/src/node_provider_auth.rs]: Fixed

**Review by yahortsaryk — APPROVED** [review]

File: pallets/ddc-clusters/src/benchmarking.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+const USER_SEED: u32 = 999666;
+const USER_SEED_2: u32 = 999555;
+
+benchmarks! {
+  where_clause { where
+		T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]> }
+
+	create_cluster {
+	  let cluster_id = ClusterId::from([1; 20]);
+		let user = account::<T::AccountId>("user", USER_SEED, 0u32);
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/benchmarking.rs:26]: Why do we have different indentations for these lines of code? Maybe the `fmt` formatting needs to be applied?

**Pull request merged**